getbouncer/cardscan-ios
# Fixing landscape/portrait bug by forcing portrait

#### Overview:
To mitigate landscape / portrait mode snafu, I initially started by mocking the UI during transition. Did not work. Then I tried to do the navigation controller overriding `supportedInterfaceOrientation` method that is all over google, the main solution. Also did not work. So the solution I came up with is `1. Lock scanViewController to portrait before transition` and `2. Do not allow orientation transitions once in scanViewController` using `appDelegate` and setting the `UIDevice.current.orientation` to portrait. 

#### To Test:
1. Make sure your device lock is not on
2. Click `Present Modally` button
3. Go in either portrait or landscape mode before clicking `To ScanViewController`
4. Rotate phone like a madman
    * Notice that view controller locks into portrait and will not rotate

#### How it works:
1. ScanViewController life cycle state stored in CardScan. 
    * Can't store in ScanViewController to access in AppDelegate so stored in CardScan
2. ScanViewController's `viewWillAppear` locks _DEVICE_ orientation to always load as portrait 
3. If any rotations are made, appDelegate ` func application(supportedInterfaceOrientationsFor window)` is called
    * Takes scanViewController life cycle state and locks _APP_ orientation to portrait.
#### Notes and Questions:
* Not ideal to make a state flag for scanViewController but it's the only way I can think of without doing nested if statements of "is it navigation controller? what is current vc? is it a scan view controller? is it tab bar controller?" . 
* `supportedInterfaceOrientation` is the range of orientations it supports, `preferredInterfaceOrientation` is a single orientation it should take within supported. 
* `supportedInterfaceOrientations` runs only if `shouldAutoRotate` is true. [Stated in first paragraph of discussion section](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621435-supportedinterfaceorientations).
